pandas/plotting/_matplotlib/core.py

```diff
@@ -240,18 +246,28 @@ def _iter_data(self, data=None, keep_index=False, fillna=None):
         if fillna is not None:
             data = data.fillna(fillna)
 
-        for col, values in data.items():
-            if keep_index is True:
-                yield col, values
-            else:
-                yield col, values.values
+        iter_data = data
+        if self.by is not None:
+            # select sub-columns based on the value of first level of MI
+            cols = data.columns.levels[0]
+            iter_data = {
```

Comment:
what are you trying to do here?

Reply:
thanks for your comment! @jreback 

This is to create a `iter_data` which is the subset of original `data` based on the unique values created by `by` argument. This is mainly to simplify the code, and avoid code duplications.

Reply:
I thing this is out of place. This `_iter_data` should be just iterating data, and not doing this extra transformation.

Not sure if it's possible to have this transformation in a different function, and call it before `_iter_data` is called.

If you need to touch this function (or even if you finally don't), would be helpful to add a docstring and typing, reading it in a review (without enough context) makes it very difficult to understand what it does now, and what it does after your change.

Reply:
thanks for your comment! @datapythonista 

I think here is kinda transformation to prepare data to be iterated. The reason to have it here is, we dont want to change `self.data` in this function, but to iterate its value. However, `df.items()` used here can only yield `column name` and `Series` tuple; but to plot for example multiple histograms in the same figure when `by` is defined, `column name` and `DataFrame` tuple are needed, and `df.items` cannot achieve this. Therefore, I took a walkaround and chose to convert the `iter_data` to `dictionary` and to be iterated so to serve plotting methods.

I added a docstring for this `_iter_data` function to make the change a bit clearer. But if you have any idea to improve this, that would be super welcome and appreciated!

thanks a lot for your review @datapythonista

Reply:
I don't think we want a function `_iter_data` to do lots of logic. The name becomes misleading, and the code much more complex.

I don't think you need this, I guess we're already plotting in different plots in some cases. Look at how they transform the data, and apply the same transformation, so this functions returns the correct output.